ES2018 implementation

bakkot opened the pull request:
Subsumes https://github.com/shapesecurity/shift-fuzzer-js/pull/26, which was blocked on fixing a bunch of upstream bugs which have now been fixed in the 2018 implementations and which I am too lazy to backport.

Protryon committed:
es2017 impl

bakkot committed:
fix some cases with async/await

bakkot committed:
es2018 implementation

michaelficarra commented on src/fuzzer-state.js:
```diff
@@ -119,6 +129,13 @@ export default class FuzzerState {
+      st.allowAwaitExpr = true;
+      st.allowAwaitIdentifier = false;
+    } else {
+      st.allowAwaitExpr = false;
```
This is the default, so no need to set this, right? Though I'm always fine with being explicit.

bakkot replied:
This state is a clone of the outer one, not fresh, so it does need to be set explicitly (for cases like `async function f(){ function g(){ await } }`).

michaelficarra commented on src/index.js:
```diff
@@ -138,19 +138,36 @@ export const fuzzArrayExpression = f =>
+  let isConcise = f.rng.nextBoolean();
+  // Because of the cover grammar we can't have an `await` identifier in the parameter list of an arrow function if we are in an async context
+  let outerForbidsAwait = !f.allowAwaitIdentifier;
+  if (!isConcise) {
```
Reverse this.

michaelficarra commented on src/index.js:
```diff
@@ -138,19 +138,36 @@ export const fuzzArrayExpression = f =>
+      f.allowAwaitIdentifier = false;
+    }
     params = fuzzFormalParameters(f, {hasStrictDirective});
+    if (outerForbidsAwait) {
```
This guard isn't necessary, but it may be okay to leave it if you think it reads easier.

bakkot replied:
Yeah, I prefer the guard, I think.

michaelficarra commented on src/index.js:
```diff
@@ -311,22 +328,34 @@ export const fuzzForInStatement = (f = new FuzzerState) => {
+  do {
+    left = f.rng.nextBoolean() ? fuzzVariableDeclaration(f, {inForInOfHead: true}) : fuzzAssignmentTarget(f);
+    // https://github.com/tc39/ecma262/issues/2034
+  } while (left.type === 'AssignmentTargetIdentifier' && left.name === 'async');
```
I don't love looping to avoid a random outcome. Can't we check for this case and replace it with something fixed?

bakkot replied:
Works for me, done.

michaelficarra commented on src/index.js:
```diff
@@ -337,26 +366,43 @@ export const fuzzFunctionBody = f =>
+  let isGenerator = false;
+  let isAsync = false;
+  if (allowProperDeclarations) {
+    let type = f.rng.nextDouble();
```
```suggestion
    let type = f.rng.nextInt(3);
```

michaelficarra commented on src/index.js:
```diff
@@ -337,26 +366,43 @@ export const fuzzFunctionBody = f =>
-  if (isGenerator) {
+  let isGenerator = false;
+  let isAsync = false;
+  let type = f.rng.nextDouble();
```
```suggestion
  let type = f.rng.nextInt(3);
```

michaelficarra commented on src/index.js:
```diff
@@ -427,7 +473,7 @@ export const fuzzLiteralNumericExpression = f =>
 export const fuzzLiteralRegExpExpression = (f = new FuzzerState, canFuzzUnicode = true) => {
   let isUnicode = canFuzzUnicode && f.rng.nextBoolean();
-  return ap(Shift.LiteralRegExpExpression, {"pattern": f => fuzzRegExpPattern(f, isUnicode), "global": f => f.rng.nextBoolean(), "ignoreCase": f => f.rng.nextBoolean(), "multiLine": f => f.rng.nextBoolean(), "sticky": f => f.rng.nextBoolean(), "unicode": f => isUnicode}, f);
+  return ap(Shift.LiteralRegExpExpression, {"pattern": f => fuzzRegExpPattern(f, isUnicode), "global": f => f.rng.nextBoolean(), "ignoreCase": f => f.rng.nextBoolean(), "multiLine": f => f.rng.nextBoolean(), "dotAll": f => f.rng.nextBoolean(), "unicode": f => isUnicode, "sticky": f => f.rng.nextBoolean()}, f);
```
It'd be nice to split these across lines.

michaelficarra commented on src/index.js:
```diff
@@ -437,20 +483,32 @@ export const fuzzMethod = (f = new FuzzerState, {isStatic = false, inClass = fal
+  let isGenerator = false;
+  let isAsync = false;
+  if (!isConstructor) {
+    let type = f.rng.nextDouble();
```
```suggestion
    let type = f.rng.nextInt(3);
```

michaelficarra commented on src/index.js:
```diff
@@ -462,13 +520,13 @@ export const fuzzNewTargetExpression = f =>
 
 export const fuzzObjectAssignmentTarget = f =>
-  ap(Shift.ObjectAssignmentTarget, {"properties": many(choose(fuzzAssignmentTargetPropertyIdentifier, fuzzAssignmentTargetPropertyProperty))}, f);
+  ap(Shift.ObjectAssignmentTarget, {"properties": many(choose(fuzzAssignmentTargetPropertyIdentifier, fuzzAssignmentTargetPropertyProperty)), "rest": opt(choose(fuzzAssignmentTargetIdentifier, choose(fuzzComputedMemberAssignmentTarget, fuzzStaticMemberAssignmentTarget)))}, f);
```
```suggestion
  ap(Shift.ObjectAssignmentTarget, {"properties": many(choose(fuzzAssignmentTargetPropertyIdentifier, fuzzAssignmentTargetPropertyProperty)), "rest": opt(choose(fuzzAssignmentTargetIdentifier, fuzzComputedMemberAssignmentTarget, fuzzStaticMemberAssignmentTarget))}, f);
```

bakkot replied:
No, having a 50% probability of being an identifier is intentional and matches the rest of this file.

michaelficarra commented on src/index.js:
```diff
@@ -746,6 +810,9 @@ export const fuzzExpression = (f = new FuzzerState, {allowIdentifierExpression =
     fuzzers = fuzzers.concat(yieldExprFuzzers);
   }
+  if (f.allowAwaitExpr) {
+    fuzzers = fuzzers.concat(fuzzAwaitExpression);
```
```suggestion
    fuzzers = fuzzers.concat([fuzzAwaitExpression]);
```

Right?

michaelficarra replied:
Also, are we trying to avoid mutation? We could just `push`.

bakkot replied:
We are trying to avoid mutation, yes. We could clone then push, but whatever.

bakkot committed:
address comments

bakkot commented:
@michaelficarra Addressed comments, though maybe this should wait for https://github.com/shapesecurity/shift-codegen-js/pull/94 - currently it's only passing by luck (although the odds of running into the problematic exact tree are pretty low).

michaelficarra commented on src/index.js:
```diff
@@ -746,6 +818,9 @@ export const fuzzExpression = (f = new FuzzerState, {allowIdentifierExpression =
   if (f.allowYieldExpr) {
     fuzzers = fuzzers.concat(yieldExprFuzzers);
```
This needs brackets too, right?

bakkot replied:
No, [it's already an array](https://github.com/shapesecurity/shift-fuzzer-js/blob/5a125892e2325e1f2b57ed41ebfa3cd6932a56f6/src/index.js#L718) (hence the plural).

bakkot committed:
bump shift-codegen dependency

bakkot committed:
prohibit \08 in directives in strict functions

bakkot committed:
bump shift-validator dependency
